### ISSUE TITLE
dependencies are checked even if buildable is False

### ISSUE THREAD

**Issue by bos** [haskell/cabal#1725]
I have a little portion of a `.cabal` file that looks like this:

```
executable judy
  main-is: Judy.hs

  buildable: False
  ghc-options: -Wall -rtsopts
  build-depends:
    base == 4.*,
    criterion,
    judy -- DELIBERATELY NOT PRESENT
```

With cabal-install 1.19.2, I get the following error:

```
    $ cabal configure
    Resolving dependencies...
*** cabal: Could not resolve dependencies:
    trying: criterion-examples-0 (user goal)
*** next goal: judy (dependency of criterion-examples-0)
    Dependency tree exhaustively searched.
```

This seems weird and non-intuitive and wrong.

**Comment by snoyberg** [comment]
I just got bit by this too. (Thanks @23Skidoo for so quickly pointing out that it was a dupe.) However, I think the "solver" label is not correct: using `runghc Setup.hs configure` can trigger this behavior too.

**Comment by mietek** [comment]
Another way to reproduce this bug:
```
cabal install -v3 --dry-run cheapskate-0.1.0.4
```
This shouldn’t pull in `wai` or `wai-extra`, as they dependencies for `cheapskate-dingus`, which isn’t buildable by default.

**Comment by dcoutts** [comment]
There's a little bit of history to this one. We had a patch (for the Cabal lib) to try to avoid non-buildable packages contributing to deps but doing so broke things in subtle ways and we had to revert it. I can't remember the details off the top of my head. The patch should be there in the git history if you look for "buildable".

**Comment by ntc2** [comment]
Me too.

**Comment by phadej** [comment]
The commit @dcoutts mentioned: https://github.com/haskell/cabal/commit/04dff517c2905e219274c56eb0f8611ac37ffc0f

**Comment by hvr** [comment]
am I right, that the current backward workaround is to use a manual flag, e.g. 

```
flag build_judy
  manual: True
  default: False

executable judy
  main-is: Judy.hs

  if !flag(build_judy)
    buildable: False
  ghc-options: -Wall -rtsopts
  if flag(build_judy)
    build-depends:
      base == 4.*,
      criterion,
      judy -- DELIBERATELY NOT PRESENT
```

?

**Comment by ntc2** [comment]
@hvr: I think the important thing in your example is that you guarded the `build-depends` with the `if flag(build_judy)`. That works! And it still works if I remove the `manual: True` on the flag. THANKS!

Complete example that actually builds:
```
name: needs-judy
version: 0
build-type: Simple
cabal-version: >= 1.2

flag build_judy
  default: False

library

executable judy
  main-is: Judy.hs

  ghc-options: -Wall -rtsopts
  if flag(build_judy)
    build-depends:
      -- DELIBERATELY NOT PRESENT
      judy
  else
    buildable: False
```
assuming a default `Setup.hs` containing:
```
import Distribution.Simple
main = defaultMain
```

**Comment by kosmikus** [comment]
The workaround that @hvr provides should indeed be ok.

I've also been working on this. Pull request (not very thoroughly tested) will follow in a few minutes.

**Comment by ezyang** [comment]
The change (that got reverted) is a0ca303448221f70b85b03a787124b22ec6d7f11

To state it more clearly, the bug is that when we flatten the package description, the set of package-global dependencies we compute blindly gobbles up all of the `build-depends`, regardless of the status of the `buildable` flag. I don't know what the subtle problem could be.

**Comment by grayjay** [comment]
This was fixed in #3039.
